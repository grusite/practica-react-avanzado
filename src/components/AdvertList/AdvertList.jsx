import React from "react";
import { withRouter } from "react-router-dom";
import { connect } from "react-redux";
import { login } from "../../actions/actions";

import Grid from "@material-ui/core/Grid";
import Typography from "@material-ui/core/Typography";
import CircularProgress from "@material-ui/core/CircularProgress";

import Advert from "../Advert/Advert";

import "../Advert/advert.css";

const mapDispatchToProps = dispatch => ({
  login: (name, surname) => dispatch(login(name, surname))
});

const mapStateToProps = state => ({
  isLoggedIn: state.user.isLoggedIn
});

class AdvertList extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      user: {
        isLoggedIn: false,
        name: "",
        surname: "",
        tags: []
      }
    };
  }

  renderAdverts = (adverts, isFetching) => {
    if (isFetching) {
      return (
        <Grid
          container
          justify="center"
          alignItems="center"
          className="card-container"
        >
          <CircularProgress
            size={80}
            thickness={3.7}
            disableShrink
            className="circular-progress"
          />
        </Grid>
      );
    }
    if (!isFetching && (!adverts || adverts.length === 0)) {
      return (
        <Grid
          container
          justify="center"
          alignItems="center"
          className="card-container"
        >
          <Typography variant="h5" component="h5">
            No existen anuncios con esos criterios de búsqueda
          </Typography>
        </Grid>
      );
    }
    return (
      <>
        <Grid
          container
          justify="center"
          alignItems="center"
          className="card-container"
        >
          <Typography variant="h5" component="h5">
            A continuación puede ver todos los anuncios disponibles
          </Typography>
        </Grid>
        <Grid
          container
          justify="space-around"
          alignItems="center"
          className="card-container"
        >
          {adverts.map(advert => (
            <Advert key={advert._id} advert={advert} />
          ))}
        </Grid>
      </>
    );
  };

  render() {
    const { adverts, isFetching } = this.props;
    return <>{adverts && this.renderAdverts(adverts, isFetching)}</>;
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withRouter(AdvertList));
