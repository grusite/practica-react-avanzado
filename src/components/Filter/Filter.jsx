import React from "react";

import Grid from "@material-ui/core/Grid";
import Typography from "@material-ui/core/Typography";
import TextField from "@material-ui/core/TextField";
import Button from "@material-ui/core/Button";
import FormGroup from "@material-ui/core/FormGroup";
import FormControlLabel from "@material-ui/core/FormControlLabel";
import Checkbox from "@material-ui/core/Checkbox";
import ShoppingBasketOutlinedIcon from "@material-ui/icons/ShoppingBasketOutlined";
import ShoppingBasketIcon from "@material-ui/icons/ShoppingBasket";
import AttachMoneyOutlinedIcon from "@material-ui/icons/AttachMoneyOutlined";
import Input from "@material-ui/core/Input";
import InputLabel from "@material-ui/core/InputLabel";
import FormControl from "@material-ui/core/FormControl";
import Select from "@material-ui/core/Select";
import InputAdornment from "@material-ui/core/InputAdornment";
import FormHelperText from "@material-ui/core/FormHelperText";

import "./filter.css";

const initialState = {
  type: "buy",
  name: "",
  description: "",
  price: 0,
  tags: [],
  tagSelected: ""
};

class Filter extends React.Component {
  constructor(props) {
    super(props);
    this.state = initialState;
    this.state.tags = this.props.tags;
    this.state.tagSelected = this.props.tagSelected;

    this.handleChange = this.handleChange.bind(this);
  }

  handleChange(event) {
    const { name, value } = event.target;

    this.setState(prevState => ({
      ...prevState,
      [name]: value
    }));
  }

  resetForm = () => {
    initialState.tags = this.state.tags;
    this.setState(initialState);
  };

  handleSubmit = () => {
    this.props.onFilterChange(this.state);
  };

  render() {
    const { type, name, description, price, tags, tagSelected } = this.state;

    return (
      <>
        <div className="bodyContainer">
          <Typography className="text" variant="h6" component="h6">
            Filtro de anuncios
          </Typography>
          <Grid container id="paper-no-material" className="paperFilter">
            <div className="adType">
              <FormGroup>
                <FormControlLabel
                  control={
                    <Checkbox
                      icon={<ShoppingBasketOutlinedIcon />}
                      checkedIcon={<ShoppingBasketIcon />}
                      value="buy"
                      name="type"
                      onChange={this.handleChange}
                      checked={type === "buy"}
                    />
                  }
                  label="Compra"
                />
                <FormControlLabel
                  control={
                    <Checkbox
                      icon={<AttachMoneyOutlinedIcon />}
                      checkedIcon={<AttachMoneyOutlinedIcon />}
                      value="sell"
                      name="type"
                      onChange={this.handleChange}
                      checked={type === "sell"}
                    />
                  }
                  label="Venta"
                />
              </FormGroup>
            </div>
            <Grid container className="inputs">
              <Grid item xs={6} sm={6} className="priceItem">
                <TextField
                  id="name"
                  name="name"
                  label="Nombre"
                  autoComplete="name"
                  value={name}
                  onChange={this.handleChange}
                />
              </Grid>
              <Grid item xs={6} sm={6} className="priceItem">
                <TextField
                  id="description"
                  name="description"
                  label="Descripción"
                  autoComplete="desc"
                  value={description}
                  onChange={this.handleChange}
                />
              </Grid>
            </Grid>
            <Grid container id="input-no-material" className="inputs">
              <Grid item xs={6} sm={6} className="priceItem">
                <FormControl fullWidth>
                  <InputLabel htmlFor="adornment-amount">Precio</InputLabel>
                  <Input
                    id="adornment-amount"
                    value={price}
                    name="price"
                    onChange={this.handleChange}
                    startAdornment={
                      <InputAdornment position="start">€</InputAdornment>
                    }
                  />
                  <FormHelperText>min-max | -max | min- </FormHelperText>
                </FormControl>
              </Grid>
              <Grid item xs={6} sm={6} className="tagItem">
                <FormControl fullWidth>
                  <InputLabel htmlFor="tag-native-simple">Tags</InputLabel>
                  <Select
                    native
                    value={tagSelected}
                    onChange={this.handleChange}
                    name="tagSelected"
                  >
                    <option value="" />
                    {tags.map((tag, index) => (
                      <option key={index} value={tag}>
                        {tag}
                      </option>
                    ))}
                  </Select>
                </FormControl>
              </Grid>
            </Grid>
          </Grid>
          <Grid item xs={12} container justify="space-around">
            <Button
              id="submit-no-material"
              type="submit"
              className="submit"
              variant="contained"
              color="primary"
              onClick={this.handleSubmit}
            >
              Filtrar
            </Button>
            <Button
              variant="contained"
              id="submit-no-material"
              color="primary"
              onClick={this.resetForm}
            >
              Restaurar
            </Button>
          </Grid>
        </div>
      </>
    );
  }
}

export default Filter;
