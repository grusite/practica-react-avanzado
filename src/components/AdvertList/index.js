import { connect } from "react-redux";
import { withRouter } from "react-router-dom";

import AdvertList from "./AdvertList";
import { login } from "../../actions/actions";

const mapDispatchToProps = dispatch => ({
  login: (name, surname, tag) => dispatch(login(name, surname, tag))
});

const mapStateToProps = state => ({
  isLoggedIn: state.user.isLoggedIn
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withRouter(AdvertList));
