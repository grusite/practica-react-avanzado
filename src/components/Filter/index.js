import { connect } from "react-redux";
import { withRouter } from "react-router-dom";

import Filter from "./Filter";
import { login } from "../../actions/actions";
import { getTags } from "../../selectors";

const mapDispatchToProps = dispatch => ({
  login: (name, surname, tag) => dispatch(login(name, surname, tag))
});

const mapStateToProps = state => ({
  tags: getTags(state)
});

export default connect(mapStateToProps, mapDispatchToProps)(withRouter(Filter));
