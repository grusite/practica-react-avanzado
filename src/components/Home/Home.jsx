import React, { useState, useEffect } from 'react'

import NavBar from '../Navbar/Navbar'
import AdvertList from '../AdvertList/AdvertList'
import Filter from '../Filter/Filter'
import './home.css'

import storage from '../../utils/storage'
const { getItem } = storage()

export default function Home(props) {
  const { advertsReducer, login, loginReducer, fetchAdverts, history } = props
  const [params, setParams] = useState(loginReducer.tag ? `tag=${loginReducer.tag}` : '')
  const [tagSelected, setTagSelected] = useState(loginReducer.tag)

  // Si no está logado le llevo a registro
  const userReminded = JSON.parse(getItem('NodePop-User'))
  const userStored = loginReducer.isLoggedIn
  if (!userReminded && !userStored) {
    history.push('/register')
  }

  // Si lo está y recarga la pagina, le vuelvo a guardar en el estado el usuario
  if (!userStored) {
    login(
      JSON.parse(getItem('NodePop-User')).name,
      JSON.parse(getItem('NodePop-User')).surname,
      JSON.parse(getItem('NodePop-User')).tag
    )
  }

  /* eslint-disable*/
  useEffect(() => {
    // filterAdverts(params).then(adverts => fetchAdverts(adverts))
    fetchAdverts(params)
  }, [params])

  const onFilterChange = state => {
    let newParam = ''
    for (let param in state) {
      if (state[param] && state[param].length !== 0 && param !== 'tags') {
        if (param === 'tagSelected') {
          setTagSelected(state[param])
          newParam += `&tag=${state[param]}`
          continue
        }

        // triquiñuela para que pueda filtrar en la api por sell or buy
        if (param === 'type') {
          if (state[param] === 'sell') {
            newParam += `&venta=true`
            continue
          }
          newParam += `&venta=false`
          continue
        }
        newParam += `&${param}=${state[param]}`
      }
    }
    setParams(newParam)
  }

  return (
    <>
      <NavBar />
      <Filter onFilterChange={onFilterChange} tagSelected={tagSelected} />
      <AdvertList adverts={advertsReducer.adverts} isFetching={advertsReducer.ui.isFetching} />
    </>
  )
}
