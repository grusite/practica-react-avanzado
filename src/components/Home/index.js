import { connect } from 'react-redux'

import Home from './Home'
import { login, fetchAdverts } from '../../actions/actions'

const mapDispatchToProps = dispatch => ({
  login: (name, surname, tag) => dispatch(login(name, surname, tag)),
  fetchAdverts: params => fetchAdverts()(dispatch, params),
})

const mapStateToProps = state => ({
  ...state,
})

export default connect(mapStateToProps, mapDispatchToProps)(Home)
