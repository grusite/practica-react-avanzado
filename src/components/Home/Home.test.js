import React from 'react'
import { shallow } from 'enzyme'

import Home from './Home'
import AdvertList from '../AdvertList/AdvertList'
import Filter from '../Filter/Filter'

describe('Home component test', () => {
  const defaultProps = {
    advertsReducer: {
      adverts: [1, 2, 3],
      ui: {
        isFetching: true,
      },
    },
    login: jest.fn(),
    loginReducer: {
      isLoggedIn: true,
    },
    fetchAdverts: jest.fn(),
    history: {
      push: jest.fn(),
    },
  }

  const render = props => shallow(<Home {...defaultProps} />)
  let wrapper

  beforeEach(() => {
    wrapper = render()
  })

  it('should render', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should render the filter component', () => {
    expect(wrapper.find(Filter).props().onFilterChange).toHaveLength(1)
  })

  it('should render the AdvertList component', () => {
    expect(wrapper.find(AdvertList).props().adverts).toHaveLength(3)
  })
})
