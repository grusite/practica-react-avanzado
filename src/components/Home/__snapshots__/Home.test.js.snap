// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component test should render 1`] = `ShallowWrapper {}`;
